// package com.java.backend.configuration;

// import lombok.NonNull;
// import org.springframework.context.annotation.Bean;
// import org.springframework.context.annotation.Configuration;
// import org.springframework.http.HttpMethod;
// import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity;
// import org.springframework.security.config.annotation.web.builders.HttpSecurity;
// import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
// import org.springframework.security.config.annotation.web.configuration.WebSecurityCustomizer;
// import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
// import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
// import org.springframework.security.crypto.password.PasswordEncoder;
// import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationConverter;
// import org.springframework.security.oauth2.server.resource.authentication.JwtGrantedAuthoritiesConverter;
// import org.springframework.security.web.SecurityFilterChain;
// import org.springframework.web.servlet.config.annotation.CorsRegistry;
// import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;

// @Configuration
// @EnableWebSecurity
// @EnableMethodSecurity
// public class SecurityConfig {

//     private static final String[] PUBLIC_ENDPOINTS_POST = {
//             "/users/register", "/auth/login", "/auth/validate-token",
//             "/users/forgot-password", "/users/confirm-account/**",
//             "/comments/{documentId}/{commentId}/like", "/users/change-password"
//     };

//     private static final String[] PUBLIC_ENDPOINTS_GET = {
//             "/users/change-password/**","/users/confirm-account/**",
//             "/auth/google-login", "/auth/google-login-success","/categories/**","/documents/**","/comments**"
//     };

//     private static final String[] PUBLIC_ENDPOINTS_PUT = {
//         "/users/change-password"
//     };

//     private final CustomJwtDecoder customJwtDecoder;

//     public SecurityConfig(CustomJwtDecoder customJwtDecoder) {
//         this.customJwtDecoder = customJwtDecoder;
//     }
//     @Bean
//     public SecurityFilterChain filterChain(HttpSecurity httpSecurity) throws Exception {
//         httpSecurity.authorizeHttpRequests(request -> request.requestMatchers(HttpMethod.POST, PUBLIC_ENDPOINTS_POST).permitAll()
//                 .requestMatchers(HttpMethod.GET, PUBLIC_ENDPOINTS_GET).permitAll()
//                 .requestMatchers(HttpMethod.PUT, PUBLIC_ENDPOINTS_PUT).permitAll()
//                 .requestMatchers(HttpMethod.DELETE,"/comments/{documentId}/{commentId}/like").permitAll()
//                 .anyRequest().authenticated()
//         );
//         httpSecurity.oauth2Login(oauth2 -> oauth2.loginPage("/auth/google-login").defaultSuccessUrl("/auth/google-login-success",true));
//         httpSecurity.oauth2ResourceServer(oauth2 -> oauth2.jwt(jwtConfigurer -> jwtConfigurer
//                         .decoder(customJwtDecoder)
//                         .jwtAuthenticationConverter(jwtAuthenticationConverter()))
//                 .authenticationEntryPoint(new JwtAuthenticationEntryPoint()));
//         httpSecurity.csrf(AbstractHttpConfigurer::disable);

//         return httpSecurity.build();
//     }

//     @Bean
//     public WebMvcConfigurer corsConfigurer() {
//         return new WebMvcConfigurer() {
//             @Override
//             public void addCorsMappings(@NonNull CorsRegistry registry) {
//                 registry.addMapping("/**")
//                         .allowedOrigins("http://localhost:8088","http://localhost:3000")
//                         .allowedMethods("GET", "POST", "PUT", "DELETE") // Allowed HTTP methods
//                         .allowedHeaders("*") // Allowed request headers
//                         .allowCredentials(true)
//                         .maxAge(3600);
//             }
//         };
//     }

//     @Bean
//     public WebSecurityCustomizer webSecurityCustomizer() {
//         return webSecurity ->
//                 webSecurity.ignoring()
//                         .requestMatchers(
//                                 "/actuator/**",
//                                 "/v3/**",
//                                 "/webjars/**",
//                                 "/swagger-ui*/*swagger-initializer.js",
//                                 "/swagger-ui*/**",
//                                 "/v3/api-docs/**"
//                         );
//     }


//     @Bean
//     JwtAuthenticationConverter jwtAuthenticationConverter() {
//         JwtGrantedAuthoritiesConverter jwtGrantedAuthoritiesConverter = new JwtGrantedAuthoritiesConverter();
//         jwtGrantedAuthoritiesConverter.setAuthorityPrefix("");

//         JwtAuthenticationConverter jwtAuthenticationConverter = new JwtAuthenticationConverter();
//         jwtAuthenticationConverter.setJwtGrantedAuthoritiesConverter(jwtGrantedAuthoritiesConverter);

//         return jwtAuthenticationConverter;
//     }

//     @Bean
//     PasswordEncoder passwordEncoder() {
//         return new BCryptPasswordEncoder(10);
//     }
// }
